src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/DictionaryLoggerScope.cs

```diff
@@ -12,16 +12,16 @@ internal class DictionaryLoggerScope
     {
         private static AsyncLocal<DictionaryLoggerScope> _value = new AsyncLocal<DictionaryLoggerScope>();
 
-        private DictionaryLoggerScope(IReadOnlyDictionary<string, object> state, DictionaryLoggerScope parent)
```

Comment:
I did try `IExternalScopeProvider` but it didn't work.
1. As you mentioned, we need the scope in the `TelemetryInitializer` for Dependencies. All telemetry items flow through the `ApplicationInsightsLogger` except dependencies. `TelemetryInitializer` is the only way I can think of to enrich dependencies with additional information.
2. IExternalScopeProvider calls an internal `BeginScope` instead of the [one](https://github.com/Azure/azure-webjobs-sdk/blob/dev/src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs#:~:text=public%20IDisposable-,BeginScope,-%3CTState%3E(TState%20state)) provided by us. We need this for `StartTelemetryIfFunctionInvocation`. `ApplicationInsightsLogger` is the only logger that uses `BeginScope`.

I did some benchmarking against IExternalScopeProvider and it's very efficient and works like magic. Unfortunately, couldn't use it here. 
The proposed changes are efficient as well, we are saving almost 99.74% allocation even with all the complexities.

Reply:
sure, I don't have a commit, but I can make a quick change and push to a branch.

Reply:
I created a branch that uses [`IExternalScopeProvider`](https://github.com/Azure/azure-webjobs-sdk/compare/roranjan/IExternalScopeProvider?expand=1). I don't see any issue with `BeginScope()` now and the only thing that's stopping us from using this is [`WebJobsTelemetryInitializer`](https://github.com/Azure/azure-webjobs-sdk/blob/roranjan/IExternalScopeProvider/src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Initializers/WebJobsTelemetryInitializer.cs#:~:text=IDictionary%3Cstring%2C%20object%3E%20scopeProps%20%3D%20new%20Dictionary%3Cstring%2C%20object%3E()%3B//DictionaryLoggerScope.GetMergedStateDictionaryOrNull()%3B).

Reply:
I did some [benchmarking ](https://github.com/RohitRanjanMS/LoggerFactoryScopeProviderBenchmarking/tree/master/LoggerFactoryScopeProviderBenchmarking) using `LoggerFactoryScopeProvider`. While the performance is almost same, memory allocation is significantly less.

<img width="614" alt="image" src="https://user-images.githubusercontent.com/90008725/226382842-adda4e03-8e15-4422-9a40-31942e1338cb.png">